File: services/search/app/trace.py
"""Trace-context helpers.

Istio's Envoy sidecars create the spans; the application's job is to read the
incoming trace id (for log correlation) and — because this service calls the
catalog downstream during reindex — forward the propagation headers so one trace
spans both services.
"""
from __future__ import annotations

from contextvars import ContextVar

# Per-request trace id, set by the middleware and read by the log formatter.
trace_id_var: ContextVar[str] = ContextVar("trace_id", default="")

# Headers Istio uses; forward these verbatim on every downstream call.
PROPAGATION_HEADERS = (
    "x-request-id",
    "traceparent",
    "tracestate",
    "x-b3-traceid",
    "x-b3-spanid",
    "x-b3-parentspanid",
    "x-b3-sampled",
    "x-b3-flags",
    "b3",
)


def trace_id_from_headers(headers) -> str:
    """Extract a trace id for log correlation, preferring W3C, then B3, then req-id."""
    tp = headers.get("traceparent", "")
    if tp:
        # traceparent = <version>-<trace-id>-<parent-id>-<flags>
        parts = tp.split("-")
        if len(parts) >= 2 and len(parts[1]) == 32:
            return parts[1]
    return headers.get("x-b3-traceid") or headers.get("x-request-id") or ""


def propagation_headers(headers) -> dict[str, str]:
    """Pick the trace-propagation headers out of an incoming request, to forward
    verbatim on a downstream call so the trace stays unbroken."""
    out: dict[str, str] = {}
    for name in PROPAGATION_HEADERS:
        val = headers.get(name)
        if val:
            out[name] = val
    return out

File: services/payment/trace.go
package main

import (
	"context"
	"log/slog"
	"net/http"
	"strings"
)

// ctxKey is an unexported type for context keys defined in this package.
type ctxKey int

const (
	traceIDKey ctxKey = iota
	routeKey
)

// propagationHeaders are the tracing/correlation headers we copy from an
// incoming request onto any outgoing request, so a single trace stitches
// together across services through the Istio mesh (Envoy + Jaeger/Tempo).
//
// payment makes no downstream calls today, but we keep the list so that any
// future hop forwards the headers — dropping them breaks the trace at this hop.
var propagationHeaders = []string{
	"x-request-id",
	"traceparent", // W3C trace context
	"tracestate",
	"x-b3-traceid", // Zipkin/B3 (Istio's default)
	"x-b3-spanid",
	"x-b3-parentspanid",
	"x-b3-sampled",
	"x-b3-flags",
	"b3",
}

// traceIDFromTraceparent extracts the 32-hex-char trace-id from a W3C
// traceparent header of the form: <version>-<trace-id>-<parent-id>-<flags>.
func traceIDFromTraceparent(tp string) string {
	parts := strings.Split(tp, "-")
	if len(parts) >= 2 && len(parts[1]) == 32 {
		return parts[1]
	}
	return ""
}

// traceIDFrom returns the best available trace identifier for log correlation,
// preferring the W3C trace-id, then B3, then the request id.
func traceIDFrom(r *http.Request) string {
	if tid := traceIDFromTraceparent(r.Header.Get("traceparent")); tid != "" {
		return tid
	}
	if tid := r.Header.Get("x-b3-traceid"); tid != "" {
		return tid
	}
	return r.Header.Get("x-request-id")
}

// propagate builds a header set to attach to outgoing requests, copying the
// trace-context headers from the incoming request.
func propagate(r *http.Request) http.Header {
	h := http.Header{}
	for _, key := range propagationHeaders {
		if v := r.Header.Get(key); v != "" {
			h.Set(key, v)
		}
	}
	return h
}

// loggerFrom returns a logger annotated with the request's trace id (if any),
// so every log line can be joined to its distributed trace in Loki/Tempo.
func loggerFrom(ctx context.Context) *slog.Logger {
	if tid, ok := ctx.Value(traceIDKey).(string); ok && tid != "" {
		return slog.With("trace_id", tid)
	}
	return slog.Default()
}
